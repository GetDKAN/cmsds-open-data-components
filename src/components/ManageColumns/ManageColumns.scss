.dkan-manage-columns-wrapper {
  .fa-columns:before {
    font-size: .75em !important;
  }
}

dialog.dkan-manage-columns-dialog {
  height: 80vh;
  width: 100%;
  .ds-c-dialog__window {
    min-width: 768px; // TODO responsive
    padding: 0;
    .ds-c-dialog__header {
      padding: 24px;
    }
  }
  .dkan-manage-columns-list {
    list-style: none;
    gap: 0;
    max-height: 350px; // TODO responsive
    margin: 0;
    padding: 0;
    overflow-y: auto;

    li {
      cursor: pointer;
    }

    .ds-c-choice-wrapper {
      margin-top: 0;
    }

    .dkan-manage-columns-reorder-button {
      background: none;
      border: none;

      &.grabbed {
        background-color: var(--color-primary-lighter);
      }
    }
  }
}

@media screen and (max-width: 1023px) {
  .dkan-manage-columns-wrapper {
    .fa-columns:before {
      font-size: 1.4em !important;
    }
  }
}

@media screen and (max-width: 767px) {
  dialog.dkan-manage-columns-dialog  {
    height: 100%;
    .ds-c-dialog__window {
      margin-top: 0;
      width: 100%;
      min-width: auto;
      height: 100%;
    }
    .ds-c-dialog__body > div:not(.ds-c-dialog__actions) {
      display: flex;
      flex-direction: column;
    }
    .dkan-manage-columns-list {
      max-height: calc(100vh - 435px); /* approx height of all other dialog elements */
    }

    .ds-c-dialog__actions {
      position:absolute;
      bottom: 0;
      background-color: white;
      padding-top: 16px;
    }
  }
}