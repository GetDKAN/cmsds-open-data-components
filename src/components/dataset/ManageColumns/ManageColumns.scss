@use "sass:list";
@use "../../../scss/modules/modules";
@use "../../../scss/modules/colormap";
@use "../../../scss/modules/functions";
@use "../../../scss/modules/variables";

// Desktop
$header-height: 64px;
$subheader-height: 56px;
$h-padding-md: 24px;
$footer-height: 80px;

// Mobile
$subheader-height-xs: 48px;
$h-padding-xs: 16px;
$footer-height-xs: 72px;

.ds-c-dialog-wrap {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0 !important;
  overflow: hidden !important;
}

.manage-col-dialog {
  &.ds-c-dialog {
    font-family: variables.$muli;
    color: colormap.$charcoal;
    padding: 0;
    z-index: 12;
    border-radius: 2px;
    width: 100%;
    max-width: 768px;
    box-shadow: 0 2px 12px 0 rgba(0, 0, 0, 0.4);

    // Manage columns sizers
    $manage-col-size: sm, md, lg, xl;
    $base-size: 40%;
    $multiplier: 10%;
    @for $i from 1 through list.length($manage-col-size) {
      $element: list.nth($manage-col-size, $i);
      &.manage-col-#{$element} {
        height: $base-size + ($multiplier * $i);
      }
    }

    > div {
      height: 100%;

      .ds-c-dialog__body {
        height: calc(100% - (#{$header-height}));

        display: flex;
        flex-direction: column;

        div[id^="dialog"][id$="__content"] {
          flex: 1;
          overflow: scroll;
        }
      }
    }

    .ds-c-dialog__header {
      display: flex;
      align-items: center;
      justify-content: space-between;
      padding: 0 $h-padding-md;
      border-bottom: 1px solid colormap.$border-color;
      min-height: $header-height;

      h1 {
        @include functions.h(3);
        & {
          margin: 0;
          font-weight: 800;
          font-size: 20px;
        }
      }

      button.ds-c-dialog__close {
        margin: 0;
        color: colormap.$charcoal;

        &:hover {
          background-color: colormap.$charcoal-5;
        }

        &::before {
          color: colormap.$charcoal;
        }
      }
    }

    .column-labels {
      @include functions.box-shadow(top);
      display: flex;
      align-items: center;
      justify-content: space-between;
      min-height: $subheader-height;
      padding: 0 $h-padding-md;
      z-index: 2;
      position: relative;
      border-bottom: 1px solid colormap.$border-color;
      background-color: colormap.$background-color;
      font-size: 14px;
      font-family: "Geometria", sans-serif;
      font-weight: 800;
    }

    #reorder-help {
      padding: .5em $h-padding-md;
      border-bottom: 1px solid colormap.$border-color;
      background-color: colormap.$background-color;
      margin: 0;
      font-size: 1em;
      white-space: normal;
    }

    .advanced-options {
      list-style: none;
      margin: 0;
      padding: 0;
      overflow-y: auto;
      height: calc(100% - #{$subheader-height});
      background-color: colormap.$background-color;

      > li[draggable] {
        padding-left: 24px;
        cursor: grab;
        display: flex;
        margin-bottom: -1px;
        box-shadow: 0 1px 0 0 colormap.$charcoal-20 inset, 0 -1px 0 0 colormap.$charcoal-20 inset;

        &:first-child {
          margin-top: -1px;
        }

        &:last-child {
          margin-bottom: 0;
        }

        // Contains label, markup controled by CMS.gov
        div:first-child {
          flex: 1;
        }

        $checkbox-left-margin: $h-padding-md;

        .ds-c-choice {
          + label.ds-c-label {
            padding: 5px 24px 5px 10px;
            margin-top: -5px;
            max-width: initial;

            &::before {
              transform: translateY(-50%);
              top: 50%;
              left: $checkbox-left-margin;
            }
          }
        }

        .manage-columns-reorder {
          border: none;
          padding: 0;
          cursor: grab;
          background: none;
          display: flex;
          align-items: center;
          box-shadow: 1px 0 0 0 colormap.$charcoal-15 inset, 0 1px 0 0 colormap.$charcoal-15 inset,
            0 -1px 0 0 colormap.$charcoal-15 inset;

          &:hover {
            background-color: colormap.$charcoal-20;
            box-shadow: 1px 0 0 0 colormap.$charcoal-15 inset,
              0 1px 0 0 colormap.$charcoal-15 inset, 0 -1px 0 0 colormap.$charcoal-15 inset;
          }

          .fa-sort {
            margin: 0 40px;
          }

          &.grabbed {
            background-color: colormap.$charcoal-20;
          }
        }
      }
    }

    // Footer
    .ds-c-dialog__actions {
      @include functions.box-shadow(bottom);
      margin: 0;
      padding: 0 $h-padding-md;
      display: flex;
      align-items: center;
      border-top: 1px solid colormap.$border-color;
      height: $footer-height;
      position: relative;
      z-index: 2;
      overflow: auto;

      .manage-col-dialog-actions {
        padding: 12px 24px;
        display: flex;
        align-items: center;
        justify-content: flex-end;
        flex: 1;
        background-color: #fff;
        white-space: nowrap;

        .select-all-choice {
          > .ds-c-choice-wrapper {
            display: flex;
            flex-grow: 1;
            justify-content: flex-end;
          }

          &.ds-c-choice--partial {
            .ds-c-label {
              &::before {
                border: 1px solid colormap.$dragonfruit;
              }

              // Indeterminate dash
              &::after {
                content: "";
                height: 3px;
                width: 16px;
                background-color: colormap.$dragonfruit;
                position: absolute;
                left: 0;
                transform: translateX(50%);
              }
            }
          }
        }
      }
    }
  }
}

@media screen and (max-width: 767px) {
  .manage-col-dialog {
    &.ds-c-dialog[class*="manage-col-"] {
      height: 100%;
      border-radius: 0;

      > div {
        .ds-c-dialog__body {
          height: calc(100% - (#{$header-height} + #{$footer-height-xs}));
        }
      }

      .ds-c-dialog__header {
        padding: 0 $h-padding-xs;

        h1 {
          @include functions.h(3);
        }
      }

      .column-labels {
        min-height: $subheader-height-xs;
        padding: 0 $h-padding-xs;
      }

      .advanced-options {
        height: calc(100% - #{$subheader-height-xs});

        > li[draggable] {
          $checkbox-left-margin: $h-padding-xs;

          .ds-c-choice {
            + label.ds-c-label {
              padding: 20px 24px 20px (44px + $checkbox-left-margin);

              &::before {
                left: $checkbox-left-margin;
              }
            }
          }

          .manage-columns-reorder {
            .fa-sort {
              margin: 0 32px;
            }
          }
        }
      }

      // Footer
      .ds-c-dialog__actions {
        padding: 0 $h-padding-xs;
        height: $footer-height-xs;
      }
    }
  }
}
