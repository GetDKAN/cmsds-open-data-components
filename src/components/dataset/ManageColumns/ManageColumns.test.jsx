import { render, screen, fireEvent, act } from '@testing-library/react'
import ManageColumns from './ManageColumns'
import FilteredDatasetResource from '../DatasetResource/FilteredDatasetResource'
import { filteredDatasetResource } from '../../../utilities/data-mocks/data-filteredDatasetResource'
import axios from 'axios'

// Mock axios
vi.mock('axios')

// Mock window.scrollTo
global.scrollTo = vi.fn()

const mockApiResponse = {
  'results': [
    {
      measure_name: 'Average Medicare spending associated with an agency\u0027s home health episodes compared with all home health episodes nationally - agency score',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'Average Medicare spending associated with an agency\u0027s home health episodes compared with all home health episodes nationally - count',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'For patients with diabetes, how often the home health team got doctor\u0027s orders, gave foot care, and taught patients about foot care',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often home health patients had new or worsened pressure ulcers',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often home health patients had to be admitted to the hospital',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often home health patients, who have had a recent hospital stay, had a preventable hospital readmission within 30 days of discharge from home health',
      measure_date_range: 'January 1, 2016 - December 31, 2018'
    },
    {
      measure_name: 'How often patients got better at bathing',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients got better at getting in and out of bed',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients got better at taking their drugs correctly by mouth',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients got better at walking or moving around',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients receiving home health care needed any urgent, unplanned care in the hospital emergency room - without being admitted to the hospital',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients remained at home within 31 days of being discharged from home health',
      measure_date_range: 'January 1, 2017 - December 31, 2018'
    },
    {
      measure_name: 'How often patients\u0027 breathing improved',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often patients\u0027 wounds improved or healed after an operation',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team began their patients\u0027 care in a timely manner',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team checked patients for depression',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team checked patients\u0027 medications and got doctor\u0027s orders for medication issues in a timely manner',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team checked patients\u0027 risk of falling',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team made sure that their patients have received a flu shot for the current flu season',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    },
    {
      measure_name: 'How often the home health team made sure that their patients have received a pneumococcal vaccine (pneumonia shot)',
      measure_date_range: 'January 1, 2018 - December 31, 2018'
    }
  ],
  count: 21,
  schema: {
    'a4eb46a4-fd19-55b0-baf4-de08c3a70abd': {
      fields: {
        measure_name: {
          type: 'text',
          mysql_type: 'text',
          description: 'Measure Name'
        },
        measure_date_range: {
          type: 'text',
          mysql_type: 'text',
          description: 'Measure Date Range'
        }
      }
    }
  },
  query: {
    keys: true,
    limit: 20,
    offset: 0,
    resources: [
      {
        id: 'a4eb46a4-fd19-55b0-baf4-de08c3a70abd',
        alias: 't'
      }
    ],
    count: true,
    results: true,
    schema: true,
    format: 'json',
    rowIds: false,
    properties: [
      'measure_name',
      'measure_date_range'
    ]
  }
}

const componentArgs = {
  datasetTitle: 'Supplier Directory Data',
  showTooltip: false
}

const renderComponent = async (args = componentArgs) => render (
  <FilteredDatasetResource
    resource={filteredDatasetResource}
  >
    <ManageColumns {...args} />
  </FilteredDatasetResource>
)

describe('ManageColumns component.', () => {
  it('Matches snapshot.', async () => {
    axios.mockResolvedValueOnce({
      data: {
        ...mockApiResponse
      }
    })
    
    let renderedManageColumns = null
    
    await act(async () => {
      renderedManageColumns = await renderComponent()
    })

    expect(renderedManageColumns.asFragment()).toMatchSnapshot()
  })

  it('The list of columns can be reordered with a keyboard and screen reader feedback is provided.', async () => {
    axios.mockResolvedValueOnce({
      data: {
        ...mockApiResponse
      }
    })
    
    // Set fake timers
    vi.useFakeTimers()

    let element = null

    await act(async () => {
      element = await renderComponent()
    })

    // Open the dialog
    fireEvent.click(screen.getByRole('button', { name: 'Manage columns - Opens in a dialog' }))
    expect(element.container.querySelector('.ds-c-dialog-wrap')).toHaveClass('open')

    // Clicking the Measure Name column button does nothing
    fireEvent.click(screen.getByRole('button', { name: 'Reorder Measure Name column' }))
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('')

    // Grab the Measure Name column reorder button
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'Enter'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('Measure Name grabbed')

    // Move the Measure Name column into the second position
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'ArrowDown'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('The rankings have been updated, Measure Name is now ranked 2 of 2')

    // Move the Measure Name column back into the first position
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'ArrowUp'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('The rankings have been updated, Measure Name is now ranked 1 of 2')

    // Drop the Measure Name column
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'Enter'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('Measure Name dropped')

    // ===

    // Grab the Measure Name column reorder button again
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'Enter'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('Measure Name grabbed')

    // Press escape key to drop the Measure Name column
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'Escape'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('Measure Name dropped')

    // Advance time by 6 seconds to clear the aria live region content
    await act(async () => {
      vi.advanceTimersByTime(6000)
    })

    // Pressing escape key on the Measure Name column reorder button does nothing when it is not grabbed
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'Escape'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('')

    // Pressing arrow up key on the Measure Name column reorder button does nothing when it is not grabbed
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'ArrowUp'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('')

    // Pressing arrow down key on the Measure Name column reorder button does nothing when it is not grabbed
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'ArrowDown'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('')

    // ===

    // Grab the Measure Name column reorder button again
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'Enter'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('Measure Name grabbed')

    // Blurring the Measure Name column reorder button drops the button
    fireEvent.blur(screen.getByRole('button', { name: 'Reorder Measure Name column' }))

    // Advance time by 6 seconds to clear the aria live region content
    await act(async () => {
      vi.advanceTimersByTime(6000)
    })

    // Attempt to move the Measure Name column into the second position but it shouldn't because it was previously blurred
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'ArrowDown'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('')

    // Reset timers
    vi.useRealTimers()
  })

  it('Clicking a column checkbox unchecks it.', async () => {
    axios.mockResolvedValueOnce({
      data: {
        ...mockApiResponse
      }
    })
    
    let element = null

    await act(async () => {
      element = await renderComponent()
    })

    // Open the dialog
    fireEvent.click(screen.getByRole('button', { name: 'Manage columns - Opens in a dialog' }))
    expect(element.container.querySelector('.ds-c-dialog-wrap')).toHaveClass('open')

    expect(element.container.querySelector('#measure_name').checked).toBe(true)

    fireEvent.click(element.container.querySelector('#measure_name'))
    expect(element.container.querySelector('#measure_name').checked).toBe(false)
  })

  it('\'Select all\' button selects/deselects all columns.', async () => {
    axios.mockResolvedValueOnce({
      data: {
        ...mockApiResponse
      }
    })
    
    let element = null

    await act(async () => {
      element = await renderComponent()
    })

    // Open the dialog
    fireEvent.click(screen.getByRole('button', { name: 'Manage columns - Opens in a dialog' }))
    expect(element.container.querySelector('.ds-c-dialog-wrap')).toHaveClass('open')

    // Unselect all
    fireEvent.click(element.container.querySelector('.select-all-choice input[type="checkbox"]'))

    Array.from(element.container.querySelectorAll('.advanced-options > li')).forEach((item) => {
      expect(item.querySelector('input[type="checkbox"]').checked).toBe(false)
    })

    // Select all
    fireEvent.click(element.container.querySelector('.select-all-choice input[type="checkbox"]'))

    Array.from(element.container.querySelectorAll('.advanced-options > li')).forEach((item) => {
      expect(item.querySelector('input[type="checkbox"]').checked).toBe(true)
    })
  })

  it('Dialog closes when columns are updated.', async () => {
    axios.mockResolvedValueOnce({
      data: {
        ...mockApiResponse
      }
    })
    
    let element = null

    await act(async () => {
      element = await renderComponent()
    })

    // Open the dialog
    fireEvent.click(screen.getByRole('button', { name: 'Manage columns - Opens in a dialog' }))
    expect(element.container.querySelector('.ds-c-dialog-wrap')).toHaveClass('open')

    // Uncheck Measure Name column
    fireEvent.click(element.container.querySelector('#measure_name'))
    expect(element.container.querySelector('#measure_name').checked).toBe(false)

    // Click Update columns button
    fireEvent.click(screen.getByRole('button', { name: 'Update columns' }))
    expect(element.container.querySelector('.ds-c-dialog-wrap')).not.toHaveClass('open')

    // ===
    
    // Reopen the dialog
    fireEvent.click(screen.getByRole('button', { name: 'Manage columns - Opens in a dialog' }))
    expect(element.container.querySelector('.ds-c-dialog-wrap')).toHaveClass('open')

    // Grab the Measure Name column reorder button
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'Enter'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('Measure Name grabbed')

    // Move the Measure Name column into the second position
    fireEvent.keyDown(screen.getByRole('button', { name: 'Reorder Measure Name column' }), {
      key: 'ArrowDown'
    })
    expect(element.container.querySelector('.sr-only.aria-live-feedback').textContent).toBe('The rankings have been updated, Measure Name is now ranked 2 of 2')

    // Click Update columns button
    fireEvent.click(screen.getByRole('button', { name: 'Update columns' }))
    expect(element.container.querySelector('.ds-c-dialog-wrap')).not.toHaveClass('open')
  })

  it('Clicking the close dialog button closes the dialog.', async () => {
    axios.mockResolvedValueOnce({
      data: {
        ...mockApiResponse
      }
    })
    
    let element = null

    await act(async () => {
      element = await renderComponent()
    })

    // Open the dialog
    fireEvent.click(screen.getByRole('button', { name: 'Manage columns - Opens in a dialog' }))
    expect(element.container.querySelector('.ds-c-dialog-wrap')).toHaveClass('open')

    // Close the dialog
    fireEvent.click(screen.getByRole('button', { name: 'Close manage columns dialog' }))
    expect(element.container.querySelector('.ds-c-dialog-wrap')).not.toHaveClass('open')
  })

  it('Dialog height is small when there are less than 5 columns.', async () => {
    axios.mockResolvedValueOnce({
      data: {
        ...mockApiResponse
      }
    })
    
    let element = null
    
    await act(async () => {
      element = await renderComponent()
    })

    expect(element.container.querySelector('.manage-col-dialog')).toHaveClass('manage-col-sm')
  })

  it('Dialog height is medium when there are more than 4 but less than 9 columns.', async () => {
    axios.mockResolvedValueOnce({
      data: {
        ...mockApiResponse,
        schema: {
          'a4eb46a4-fd19-55b0-baf4-de08c3a70abd': {
            fields: {
              test_1: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 1'
              },
              test_2: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 2'
              },
              test_3: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 3'
              },
              test_4: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 4'
              },
              test_5: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 5'
              },
              test_6: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 6'
              },
              test_7: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 7'
              },
              test_8: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 8'
              }
            }
          }
        }
      }
    })

    let element = null
    
    await act(async () => {
      element = await renderComponent()
    })

    expect(element.container.querySelector('.manage-col-dialog')).toHaveClass('manage-col-md')
  })

  it('Dialog height is large when there are more than 8 but less than 13 columns.', async () => {
    axios.mockResolvedValueOnce({
      data: {
        ...mockApiResponse,
        schema: {
          'a4eb46a4-fd19-55b0-baf4-de08c3a70abd': {
            fields: {
              test_1: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 1'
              },
              test_2: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 2'
              },
              test_3: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 3'
              },
              test_4: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 4'
              },
              test_5: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 5'
              },
              test_6: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 6'
              },
              test_7: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 7'
              },
              test_8: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 8'
              },
              test_9: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 9'
              }
            }
          }
        }
      }
    })

    let element = null
    
    await act(async () => {
      element = await renderComponent()
    })

    expect(element.container.querySelector('.manage-col-dialog')).toHaveClass('manage-col-lg')
  })

  it('Dialog height is extra-large when there are more than 13 columns.', async () => {
    axios.mockResolvedValueOnce({
      data: {
        ...mockApiResponse,
        schema: {
          'a4eb46a4-fd19-55b0-baf4-de08c3a70abd': {
            fields: {
              test_1: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 1'
              },
              test_2: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 2'
              },
              test_3: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 3'
              },
              test_4: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 4'
              },
              test_5: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 5'
              },
              test_6: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 6'
              },
              test_7: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 7'
              },
              test_8: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 8'
              },
              test_9: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 9'
              },
              test_10: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 10'
              },
              test_11: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 11'
              },
              test_12: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 12'
              },
              test_13: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 13'
              },
              test_14: {
                type: 'text',
                mysql_type: 'text',
                description: 'Test 14'
              }
            }
          }
        }
      }
    })

    let element = null
    
    await act(async () => {
      element = await renderComponent()
    })

    expect(element.container.querySelector('.manage-col-dialog')).toHaveClass('manage-col-xl')
  })
})
