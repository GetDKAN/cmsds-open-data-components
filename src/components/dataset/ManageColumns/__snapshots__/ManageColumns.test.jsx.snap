// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ManageColumns component. > Matches snapshot. 1`] = `
<DocumentFragment>
  <div
    class="table-control-item manage-columns-button-container"
  >
    <span
      class="tooltip-trigger"
    >
      <button
        aria-label="Manage columns - Opens in a dialog"
        class="dataset-button"
      >
        <svg
          aria-label="columns"
          class="fa fa-columns"
          focusable="false"
          height="16"
          role="img"
          viewBox="0 0 512 512"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M464 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h416c26.51 0 48-21.49 48-48V80c0-26.51-21.49-48-48-48zM232 432H54a6 6 0 0 1-6-6V112h184v320zm226 0H280V112h184v314a6 6 0 0 1-6 6z"
            fill="#0c2499"
          />
        </svg>
        <span
          class="dataset-button-text"
        >
          Manage columns
        </span>
      </button>
    </span>
    <div
      class="ds-c-dialog-wrap"
    >
      <dialog
        aria-labelledby="dialog--1__heading"
        class="ds-c-dialog manage-col-dialog manage-col-sm"
        id="dialog--1"
        role="dialog"
        tabindex="-1"
      >
        <div
          class="ds-c-dialog__window"
        >
          <div
            class="ds-c-dialog__header"
          >
            <h2
              class="ds-c-dialog__heading"
              id="dialog--1__heading"
            >
              Manage columns
            </h2>
            <button
              aria-label="Close manage columns dialog"
              class="ds-c-close-button ds-c-dialog__close"
              id="dialog--1__close"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="ds-c-icon ds-c-icon--close ds-c-icon--close-thin "
                id="dialog--1__close__icon"
                viewBox="-2 -2 18 18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 13.0332964L13.0332964 0M13.0332964 13.0332964L0 0"
                  fill="none"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
          <div
            class="ds-c-dialog__body"
          >
            <div
              id="dialog--1__content"
            >
              <div
                class="column-labels"
              >
                <span>
                  Display column
                </span>
                <span>
                  Reorder
                </span>
              </div>
              <p
                id="reorder-help"
              >
                Activate the reorder button and use the arrow keys to reorder the list or use your mouse to drag/reorder. Press escape to cancel the reordering.
              </p>
              <ul
                class="advanced-options"
              >
                <li
                  draggable="true"
                  style="opacity: 1;"
                >
                  <div>
                    <div
                      class="ds-c-choice-wrapper"
                    >
                      <input
                        aria-describedby="measure_name__error"
                        checked=""
                        class="ds-c-choice"
                        id="measure_name"
                        name="Measure Name"
                        style="cursor: pointer;"
                        title="Toggle Column Visible"
                        type="checkbox"
                        value=""
                      />
                      <label
                        class="ds-c-label"
                        for="measure_name"
                        id="measure_name__label"
                      >
                        Measure Name
                      </label>
                      <p
                        aria-atomic="true"
                        aria-live="assertive"
                        class="ds-c-inline-error"
                        id="measure_name__error"
                      />
                    </div>
                  </div>
                  <button
                    aria-describedby="reorder-help"
                    aria-label="Reorder Measure Name column"
                    class="manage-columns-reorder"
                    type="button"
                  >
                    <svg
                      aria-label="sort"
                      class="fa fa-sort"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 320 512"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                        fill="#323A45"
                      />
                    </svg>
                  </button>
                </li>
                <li
                  draggable="true"
                  style="opacity: 1;"
                >
                  <div>
                    <div
                      class="ds-c-choice-wrapper"
                    >
                      <input
                        aria-describedby="measure_date_range__error"
                        checked=""
                        class="ds-c-choice"
                        id="measure_date_range"
                        name="Measure Date Range"
                        style="cursor: pointer;"
                        title="Toggle Column Visible"
                        type="checkbox"
                        value=""
                      />
                      <label
                        class="ds-c-label"
                        for="measure_date_range"
                        id="measure_date_range__label"
                      >
                        Measure Date Range
                      </label>
                      <p
                        aria-atomic="true"
                        aria-live="assertive"
                        class="ds-c-inline-error"
                        id="measure_date_range__error"
                      />
                    </div>
                  </div>
                  <button
                    aria-describedby="reorder-help"
                    aria-label="Reorder Measure Date Range column"
                    class="manage-columns-reorder"
                    type="button"
                  >
                    <svg
                      aria-label="sort"
                      class="fa fa-sort"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 320 512"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                        fill="#323A45"
                      />
                    </svg>
                  </button>
                </li>
              </ul>
              <div
                aria-atomic="true"
                aria-live="assertive"
                class="sr-only aria-live-feedback"
              />
            </div>
            <div
              class="ds-c-dialog__actions"
            >
              <div
                class="manage-col-dialog-actions"
              >
                <div
                  class="select-all-choice"
                >
                  <div
                    class="ds-c-choice-wrapper"
                  >
                    <input
                      aria-describedby="choice--2__error"
                      checked=""
                      class="ds-c-choice"
                      id="choice--2"
                      name=""
                      type="checkbox"
                      value=""
                    />
                    <label
                      class="ds-c-label"
                      for="choice--2"
                      id="choice--2__label"
                    >
                      Select all
                    </label>
                    <p
                      aria-atomic="true"
                      aria-live="assertive"
                      class="ds-c-inline-error"
                      id="choice--2__error"
                    />
                  </div>
                </div>
                <button
                  class="btn blueberry"
                >
                  Update columns
                </button>
              </div>
            </div>
          </div>
        </div>
      </dialog>
    </div>
  </div>
</DocumentFragment>
`;
